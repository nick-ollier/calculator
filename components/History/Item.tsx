import { formatHistoryExpression } from '@utils';

export interface ItemProps {
    value: string;
    expression: string;
}
const Item: React.FC<ItemProps> = ({ value, expression }) => (
    <div
        role='presentation'
        aria-label={`${expression}=${value}`}
        className='mb-4 text-right cursor-pointer'
        onClick={() => navigator.clipboard.writeText(value)}
    >
        <p className='text-sm'>{formatHistoryExpression(expression)}</p>
        <p className='text-2xl font-bold'>{value}</p>
    </div>
);

export default Item;
