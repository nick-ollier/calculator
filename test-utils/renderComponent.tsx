import CalculatorProvider from '@context/CalculatorContext';
import { render } from '@testing-library/react';

const renderWrapper = (ui: React.ReactElement) => {
    const WithProviders = ({ children }) => {
        return <CalculatorProvider>{children}</CalculatorProvider>;
    };

    return render(ui, { wrapper: WithProviders });
};

export * from '@testing-library/react';
export { renderWrapper as render };
