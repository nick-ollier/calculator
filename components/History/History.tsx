import { motion, AnimatePresence } from 'framer-motion';
import classnames from 'classnames';
import Item from '@components/History/Item';

export interface HistoryProps {
    displayHistory: boolean;
    history: {
        id: string;
        expression: string;
        value: string;
    }[];
    clearHistory: () => void;
    closeHistory: () => void;
}

const History: React.FC<HistoryProps> = ({
    displayHistory,
    history,
    clearHistory,
    closeHistory
}) => (
    <AnimatePresence>
        {displayHistory ? (
            <motion.div
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore: Motion Dev note: Awaiting Fix
                positionTransition
                initial={{
                    opacity: 0,
                    y: 10,
                    height: '80%',
                    width: '80%',
                    top: '10%',
                    left: '10%',
                    position: 'absolute',
                    display: 'flex',
                    justifyContent: 'center',
                    alignItems: 'center',
                    zIndex: 10
                }}
                animate={{ opacity: 1, y: 0 }}
                exit={{ opacity: 0, y: 10, transition: { duration: 0.25 } }}
            >
                <div
                    className={classnames(
                        'h-full w-full p-5 bg-white dark:bg-gray-500 bg-opacity-90 dark:bg-opacity-95 flex flex-col text-gray-500 dark:text-gray-100 rounded-xl'
                    )}
                    data-testid='history'
                >
                    <strong className='border-gray-200 border-b mb-5 flex justify-between items-center'>
                        History
                        <button
                            type='button'
                            onClick={closeHistory}
                            className='text-3xl min-w-btn-p text-right'
                            aria-label='Hide History'
                        >
                            &times;
                        </button>
                    </strong>

                    <div className='flex-grow overflow-scroll'>
                        {history.map(({ id, expression, value }) => (
                            <Item
                                key={id}
                                value={value}
                                expression={expression}
                            />
                        ))}
                    </div>

                    <strong className='border-gray-200 border-t mt-5 flex justify-between items-center'>
                        Clear
                        <button
                            type='button'
                            onClick={clearHistory}
                            className='text-3xl min-w-btn-p text-right'
                            aria-label='Clear History'
                        >
                            &minus;
                        </button>
                    </strong>
                </div>
            </motion.div>
        ) : null}
    </AnimatePresence>
);

export default History;
