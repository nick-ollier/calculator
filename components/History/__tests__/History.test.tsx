import { render, screen } from '@testUtils/renderComponent';
import userEvent from '@testing-library/user-event';
import History, { HistoryProps } from '@components/History/History';

const mockClearHistory = jest.fn();
const mockCloseHistory = jest.fn();
let mockHistoryData = [{ id: '12345', expression: '1+1', value: '2' }];

const renderComponent = (props?: Partial<HistoryProps>) => {
    const utils = render(
        <History
            displayHistory
            history={mockHistoryData}
            clearHistory={mockClearHistory}
            closeHistory={mockCloseHistory}
            {...props}
        />
    );

    return {
        ...utils
    };
};

describe('component: history History', () => {
    it('renders', () => {
        const { container } = renderComponent();
        expect(container);
    });

    it('matches snapshot', () => {
        const { asFragment } = renderComponent();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should fire closeHistory callback on click of close icon', () => {
        renderComponent();

        const closeHistoryButton = screen.getByRole('button', {
            name: /hide history/i
        });
        userEvent.click(closeHistoryButton);

        expect(mockCloseHistory).toHaveBeenCalled();
    });

    it('should fire clearHistory callback on click of close icon', () => {
        renderComponent();

        const clearHistoryButton = screen.getByRole('button', {
            name: /clear history/i
        });
        userEvent.click(clearHistoryButton);

        expect(mockCloseHistory).toHaveBeenCalled();
    });

    it('should render correct number of history items', () => {
        mockHistoryData = [
            { id: '1', expression: '1+1', value: '2' },
            { id: '2', expression: '2+3', value: '5' },
            { id: '3', expression: '20-22', value: '-2' }
        ];
        renderComponent({ history: mockHistoryData });

        const historyItems = screen.getAllByRole('presentation');

        expect(historyItems).toHaveLength(mockHistoryData.length);
    });

    it("shouldn't render history when displayHistory is false", () => {
        renderComponent({ displayHistory: false });

        const history = screen.queryByTestId('history');

        expect(history).not.toBeInTheDocument();
    });
});
