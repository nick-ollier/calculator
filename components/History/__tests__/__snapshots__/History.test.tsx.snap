// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: history History matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="opacity: 0; height: 80%; width: 80%; top: 10%; left: 10%; position: absolute; display: flex; justify-content: center; align-items: center; z-index: 10; transform: translateY(10px) translateZ(0);"
  >
    <div
      class="h-full w-full p-5 bg-white dark:bg-gray-500 bg-opacity-90 dark:bg-opacity-95 flex flex-col text-gray-500 dark:text-gray-100 rounded-xl"
      data-testid="history"
    >
      <strong
        class="border-gray-200 border-b mb-5 flex justify-between items-center"
      >
        History
        <button
          aria-label="Hide History"
          class="text-3xl min-w-btn-p text-right"
          type="button"
        >
          ×
        </button>
      </strong>
      <div
        class="flex-grow overflow-scroll"
      >
        <div
          aria-label="1+1=2"
          class="mb-4 text-right cursor-pointer"
          role="presentation"
        >
          <p
            class="text-sm"
          >
            1 + 1 =
          </p>
          <p
            class="text-2xl font-bold"
          >
            2
          </p>
        </div>
      </div>
      <strong
        class="border-gray-200 border-t mt-5 flex justify-between items-center"
      >
        Clear
        <button
          aria-label="Clear History"
          class="text-3xl min-w-btn-p text-right"
          type="button"
        >
          −
        </button>
      </strong>
    </div>
  </div>
</DocumentFragment>
`;
